sophia/src/triple/stream/_error.rs

```diff
@@ -0,0 +1,214 @@
+use std::convert::Infallible;
+use std::error::Error;
+use std::fmt;
+
+/// A error that is raised by functions that move fallible `Source`s into
+/// fallible `Sinks`.
+///
+/// In case this error is raised it can be matched to investigate if `Source`
+/// or `Sink` failed.
+///
+/// The `std::error::Error::source()` method is implemented accordingly.
+///
+/// # ToDo
+///
+/// For a proper use in multi-threaded environments the trait bounds should be
+/// supplied with `+ Send + Sync`. However, such bounds wound make `StreamError`
+/// incompatible with `sophia`'s whole error-handling. This should be resolved
+/// when the error-handling is completly refactored
+/// ([tracking issue](https://github.com/pchampin/sophia_rs/issues/8)).
+#[derive(Debug)]
+pub enum StreamError<SourceErr, SinkErr>
+where
+    SourceErr: 'static + Error,
+    SinkErr: 'static + Error,
+{
+    Source(SourceErr),
+    Sink(SinkErr),
+}
+
+use self::StreamError::*;
+
+impl<SourceErr, SinkErr> StreamError<SourceErr, SinkErr>
+where
+    SourceErr: 'static + Error,
+    SinkErr: 'static + Error,
+{
+	/// Build a `StreamError` from an error of the `Source`.
+    pub fn from_source_err(se: SourceErr) -> Self {
+        Source(se)
+	}
+	/// Build a `StreamError` from an error of the `Sink`.
+    pub fn from_sink_err(se: SinkErr) -> Self {
+        Sink(se)
+    }
+    /// Checks if `StreamError` was raised by the `Source`.
+    pub fn is_source_err(&self) -> bool {
+        if let Source(_) = self {
+            true
+        } else {
+            false
+        }
+    }
+    /// Checks if `StreamError` was raised by the `Sink`.
+    pub fn is_sink_err(&self) -> bool {
+        if let Sink(_) = self {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+impl<SourceErr, SinkErr> StreamError<SourceErr, SinkErr>
+where
+    SourceErr: 'static + Error,
+    SinkErr: 'static + Error + Into<SourceErr>,
+{
+	/// Converts the ``StreamError` into its `SourceError`.
+	/// 
+	/// _Note_ this also works if `SinkError` is `Infallible`.
+    pub fn into_source_err(self) -> SourceErr {
+        match self {
+            Source(err) => err,
+            Sink(err) => err.into(),
+        }
+    }
+}
```

Comment:
I think this impl and all the following similar ones could be replaced by a *single* method, in the main impl block, that would look like this:
```rust
    pub fn unwrap_into<Err>(self) -> Err where
        Err: From<SourceErr> + From<SinkErr>
    {
        match self {
            SourceError(err) => err.into(),
            SinkError(err) => err.into(),
        }
    }
````
This would make the API of `StreamError` more simple and homogeneous.

Furthermore, the naming of the methods is misleading. At first I thought that `into_source_err` was meant to change the Sink variant into the Source variant, which I thought was odd...

Reply:
I prefer to call the function `into_inner()`. For me 'unwrap' always signals some operation that might _panic_ if not called properly. Which this function clearly not does.

Reply:
Granted, `unwrap` was not the best choice.
Shall I be over-pendantic and suggest `inner_into`, because we are not exactly converting the StreamError into its inner error, we are rather converting the inner error into potentially something else...